### PR TITLE
Add GetRollupCaps API to high level rest client

### PR DESCRIPTION
**Pull request by polyfractal**
This is based on @tlrx's branch in #32703, so a review should probably hold off until that PR is merged so that this diff is a little cleaner.

The GetRollupIndexCaps API uses basically the same serialization of objects with a slightly different endpoint, but this PR was getting big enough already so I'll do that one in a followup.

Related #29827

### CODE REVIEW

**Comment by hub-cap** [comment]
hi @polyfractal, We decided to split the request and response from the ones used in server/x-pack, so they do not overlap. Please update your PR such that you create new `request implements Validatable` & `generic response class` in the high level rest client's `org.elasticsearch.client.nameOfYourClient`. Example: `o.e.client.watcher`, `o.e.client.cluster`. Please revert any work you did to modify the existing request/response in x-pack or server.

**Comment by tlrx** [comment]
@polyfractal Is it ready to review?

**Comment by polyfractal** [comment]
@tlrx sorry, not quite yet.  After pushing I realized the client versions of the clases still extend Writeable when they don't need too.  Removed those and need to refactor a few tests because of it.

I'm off today but will finish it up first thing on Monday :)

**Comment by polyfractal** [comment]
@tlrx Ok, think this is ready for a look now.  Sorry for the delay, was fighting tests :)

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RollupClient.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RollupClient.java]: Nit: doc not up to date

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsRequest.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsRequest.java]: Can be final too

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsRequest.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsRequest.java]: Usually static are placed first at the beginning of the class

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsRequest.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsRequest.java]: You don't need this, this is the default implementation

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java]: Could be final

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java]: Maybe convert `jobs` to an unmodifiable map?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java]: deepEquals instead?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java]: Both can be final

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java]: You should create a copy of the list, then sort it, then make it unmodifiable

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java]: extra line

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: can be final  too

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: Can you use anonymous blocks to indent the construction of the XContent, like:
```
builder.startObject();
{
        builder.field(JOB_ID.getPreferredName(), jobID);
}
builder.endObject();
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: + indexPattern

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: static first

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: can be final

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: Looks like it's a ToXContentFragment, not ToXContentObject?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java
Comment (tlrx) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java]: extra line

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java
Comment (tlrx) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java]: Usually we use bullet points like `// <1>` to describe lines of code, would it make sense here too?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java
Comment (tlrx) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java]: Does the test rely on the existence of the rollup index? If so we need to wait for its creation

File: client/rest-high-level/src/test/java/org/elasticsearch/client/rollup/GetRollupCapsRequestTests.java
Comment (tlrx) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/rollup/GetRollupCapsRequestTests.java]: extra lines

**Review by tlrx — COMMENTED** [review]
I left a first bunch of comments, mostly cosmetics

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (polyfractal) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: good catch thanks

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java
Comment (polyfractal) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/GetRollupCapsResponse.java]: I believe deepEquals is only needed for arrays (jobs is a map)?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java
Comment (polyfractal) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java]: I'll add a check just in case, but it shouldn't strictly be necessary.  The PutJob API won't return until the index has been created, and the CreateIndex API doesn't return successfully until the CS is updated and the primaries are allocated.

The caps info is stored in the metadata of an index, so as long as the PutJob API returns successfully it should be there.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java
Comment (polyfractal) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/RollupDocumentationIT.java]: So I went with inline comments because it's a longish snippet, and all the bullets would be located at the bottom under the snippet.  It felt better to describe as the user reads the code, rather than jumping back/forth to the bottom of the snippet each time?

I could try breaking the snippet up into shorter blocks, but that may make it more confusing since it's dealing with the same response object.

**Comment by polyfractal** [comment]
Tidied up the various style issues, and applied similar corrections to some of the other classes :)

**Comment by hub-cap** [comment]
this PR is way too big for me to review. if @tlrx is happy, then im happy. Im just going to check the high level rest items and skip all the pojos. next time it would be rad to separate all the pojos from the hlrc work.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RollupClient.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RollupClient.java]: any reason these two methods have drastically different indentation? one of them is a newline every param and the Async below is just 2 lines.

File: docs/java-rest/high-level/x-pack/rollup/put_job.asciidoc
Comment (hub-cap) [inline docs/java-rest/high-level/x-pack/rollup/put_job.asciidoc]: whats up this file? is it meant to be added?

**Review by hub-cap — COMMENTED** [review]
based on the size of this, it might make sense to break it up into > 1 PR. I cant approve it as is, because its too much to review. If @tlrx is happy, then thats good enuf for me!

**Review by hub-cap — COMMENTED** [review]
based on the size of this, it might make sense to break it up into > 1 PR. I cant approve it as is, because its too much to review. If @tlrx is happy, then thats good enuf for me!

**Review by tlrx — APPROVED** [review]
LGTM - I left some minor comments that can be addressed without another review

There still two important things: the Put Rollup documentation that does not belong to this PR and a doc change to make it compatible with #34125

**Comment by polyfractal** [comment]
Will continue the style fixes, and see if it can be split into >1 PR... I suspect that's going to be hard.  

The current structure of the client means we have to do a lot of semi-duplication, and a lot of moving things around, just to make the client bits work.  E.g. all the request/response objects need to be sorta duplicated, minus the bits that are only used internal to the server.  In the case of the Caps API, that's a lot of objects since it's a deeply-nested response object.  Plus their associated response parsers, which aren't used on the Server side.

There are some smaller changes I could probably make to Rollup in isolation (some changes to make a few objects immutable) but honestly I don't think that will help much.

Re: docs... I meant to _move_ the PutRollupJob docs, not sure why it copied instead.  The organization of the docs appears to be under `/java-rest/high-level/<plugin>/`.  But it appears that PutJob is under `/java-rest/high-level/x-pack/rollup/`... I'm assuming that was from before the recent HLRC restructuring.  In any case, it seems like the x-pack version should be deleted and moved to the other location to follow how other plugins are doing it (ML, etc)

**Comment by hub-cap** [comment]
RE docs: it should not have an x-pack folder

RE splitting, dont worry about it. if @tlrx is happy, so am I. The high points of the HLRC are ok, and assuming you fix the docs as he pointed out, :shipit:

**Comment by tlrx** [comment]
@polyfractal Once docs and the few style issues fixed I'll be happy to review this one more time

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java
Comment (polyfractal) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java]: It is... but isn't it better to sort once in the ctor instead of each time toXContent is called?  Since jobcaps is final and unmodifiable, we know it will only happen once and won't ever become unsorted.

Comment (polyfractal) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java]: (also note iirc it's mainly there to make testing from yaml rest tests easier)

**Comment by polyfractal** [comment]
Jenkins test this please

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java
Comment (tlrx) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java]: Ok, let's keep the sorting in the ctor. I wanted to avoid the sort to be done each time the object is instantiated but this is a HLRC specific object so it will be done just one time

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollupJobCaps.java]: You can use Objects.equals() here

File: client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/rollup/RollableIndexCaps.java]: You can use Objects.equals() here

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollableIndexCaps.java
Comment (tlrx) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollableIndexCaps.java]: You can use Objects.equals() here

**Review by tlrx — APPROVED** [review]
LGTM, left minor comments. If CI is happy I'm happy

**Comment by tlrx** [comment]
Thanks @polyfractal :tada:

**Pull request merged**
